File: slaspec/instructions/instr16/NOP16.py
from slaspec.instructions.core import Instruction, InstructionFamily16
from slaspec.instructions.pattern import FieldTemplate as FT, Mask


class NOP16Family(InstructionFamily16):
    def __init__(self) -> None:
        super().__init__()
        self.name = "NOP16"
        self.desc = "16-bit Slot Nop"
        self.prefix = "nop"

        self.splitField("", [FT("sig", Mask(0x0000), 16)])

        # Instruction
        NOPInstrNOP16(self)


class NOPInstrNOP16(Instruction):
    def __init__(self, family: NOP16Family) -> None:
        super().__init__(family)
        self.name = "NOP"

File: src/slaspec/instructions/instr16/nop16.rs
use crate::slaspec::instructions::{
    core::{InstrBuilder, InstrFactory, InstrFamilyBuilder},
    pattern::{FieldType, ProtoField, ProtoPattern},
};

pub fn instr_fam() -> InstrFamilyBuilder {
    let mut ifam = InstrFamilyBuilder::new_16(
        "NOP16",
        "16-bit Slot Nop",
        "nop",
        ProtoPattern {
            fields: vec![ProtoField::new("sig", FieldType::Mask(0x0000), 16)],
        },
    );

    ifam.add_instrs(&NOPFactory());

    ifam
}

struct NOPFactory();

impl InstrFactory for NOPFactory {
    fn build_instrs(&self, ifam: &InstrFamilyBuilder) -> Vec<InstrBuilder> {
        vec![InstrBuilder::new(ifam).name("NOP")]
    }
}
